modin/backends/pandas/query_compiler.py

```diff
@@ -2525,6 +2525,84 @@ def __convert_by(by):
 
         return unstacked
 
+    def pivot_table(
+        self,
+        index,
+        values,
+        columns,
+        aggfunc,
+        fill_value,
+        margins,
+        dropna,
+        margins_name,
+        observed,
+    ):
+        ErrorMessage.missmatch_with_pandas(
+            operation="pivot_table",
+            message="Order of columns could be different from pandas",
+        )
+
+        from pandas.core.reshape.pivot import _convert_by
+
+        def __convert_by(by):
+            if isinstance(by, pandas.Index):
+                return list(by)
+            return _convert_by(by)
+
+        index, columns, values = map(__convert_by, [index, columns, values])
+
+        unique_keys = np.unique(index + columns)
+        unique_values = np.unique(values)
+
+        if len(values):
+            to_group = self.getitem_column_array(unique_values)
+        else:
+            to_group = self.drop(columns=unique_keys)
+
+        keys_columns = self.getitem_column_array(unique_keys)
+
+        def applyier(df, other):
+            concated = pandas.concat([df, other], axis=1, copy=False)
+            result = concated.pivot_table(
+                index=index,
+                values=values if len(values) > 0 else None,
+                columns=columns,
+                aggfunc=aggfunc,
+                fill_value=fill_value,
+                margins=margins,
+                dropna=dropna,
+                margins_name=margins_name,
+                observed=observed,
+            )
+
+            # in that case Pandas transposes the result of `pivot_table`,
+            # transposing it back to be consistent with column axis values along
+            # different partitions
+            if len(index) == 0 and len(columns) > 0:
+                result = result.T
+
+            return result
+
+        result = self.__constructor__(
+            to_group._modin_frame.broadcast_apply_full_axis(
```

Comment:
Can we convert `keys_columns` to pandas and apply `applyier` on df and other?

Reply:
I mean the approach which is used in `merge` implementation for example.

Reply:
I'm not sure, but I think for some engines it could be slower to do `.to_pandas` for the frame rather than broadcasting it. I'm realizing that there is no difference in the shape of the broadcasted frame, which still be N full columns. 

The benefit in broadcasting here, is that we're passing to the function not serialized objects, [but its `oid`](https://github.com/dchigarev/modin/blob/1f8ee522999f7a3773e0b42ecc8b3cd56ef54137/modin/engines/ray/pandas_on_ray/frame/axis_partition.py#L22-L27) (in case of `Ray`, other engines may have their own names for that). So it's engines responsobility to serialize objects in the most efficient way, when their executing a function.

I don't have any performance measurements on it, and not closely familiar with how `Ray` and other engines obtains futures/objects ids, is this beneficial optimization or not. Let's ask @devin-petersohn since he's more competent in that question

Reply:
It would be interesting to look at performance measurement these approaches. @dchigarev , can you do some measurements?

Reply:
Here are the results of measurements for 200,000 rows DataFrame. It seems that broadcasting is significantly faster rather than `to_pandas()`
```
MEASURING single-partition columns; df shape is (200000, 255)
to_pandas 2.890704656019807
broadcasting 0.8330112155526876


MEASURING multi-partition columns; df shape is (200000, 253)
to_pandas 4.838804990053177
broadcasting 1.0249458067119122
```

<details><summary>script</summary>


```python
import numpy as np
import os
import pandas
from timeit import default_timer as timer


def generate_data_file(
    filename=None,
    nrows=200000,
    ncols=256,
    multiline_rows=False,
    force=False,
    rand_low=-100,
    rang_high=100,
    log=True,
):
    if filename is None:
        filename = os.path.abspath(
            f"int_dataset-{nrows},{ncols},{rand_low},{rang_high},{multiline_rows}.csv"
        )

    print("generating", filename)

    if os.path.exists(filename) and not force:
        return filename

    data = {
        f"col{i}": np.concatenate(
            [
                np.concatenate(
                    [
                        ["some\nvery very very\nlong string\nwith many multilines"],
                        np.random.randint(rand_low, rang_high, 9),
                    ]
                )
                for _ in np.arange(ncols // 10)
            ]
        )
        if (i % 10 == 0 and multiline_rows)
        else list(np.random.randint(rand_low, rang_high, nrows // 10)) * 10
        for i in np.arange(ncols)
    }
    print("dict generated!")
    df = pandas.DataFrame(data)
    print("dataframe created!")
    df.to_csv(filename)
    print("csv ready!")
    return filename


def measure_time(md_df, columns, test_label):
    subframe = md_df[columns]
    md_df = md_df.drop(columns=columns)

    print(f"\n\nMEASURING {test_label}; df shape is {md_df.shape}")

    t1 = timer()
    pd_subframe = subframe._to_pandas()

    def single(df):
        pd_subframe.max()
        df.max()
        return df

    result = pd.DataFrame(
        query_compiler=PandasQueryCompiler(
            md_df._query_compiler._modin_frame._apply_full_axis(0, single)
        )
    )
    repr(result)
    print("to_pandas", timer() - t1)

    t1 = timer()

    def single(df, other):
        other.max()
        df.max()
        return df

    result = pd.DataFrame(
        query_compiler=PandasQueryCompiler(
            md_df._query_compiler._modin_frame.broadcast_apply_full_axis(
                0, single, other=subframe._query_compiler._modin_frame
            )
        )
    )
    repr(result)
    print("broadcasting", timer() - t1)


if __name__ == "__main__":
    import modin.pandas as pd
    from modin.backends import PandasQueryCompiler
    from modin.pandas.test.utils import df_equals, test_data
    import pandas
    import numpy as np

    filename = generate_data_file()

    md_df = pd.read_csv(filename)

    measure_time(md_df, md_df.columns[:2], test_label="single-partition columns")
    measure_time(
        md_df,
        [*md_df.columns[:2], *md_df.columns[-2:]],
        test_label="multi-partition columns",
    )

```



</details>

Reply:
@dchigarev , thanks! It's great.